src/player.cpp

```diff
@@ -2098,8 +2098,32 @@ Item* Player::getCorpse(Creature* lastHitCreature, Creature* mostDamageCreature)
 	Item* corpse = Creature::getCorpse(lastHitCreature, mostDamageCreature);
 	if (corpse && corpse->getContainer()) {
 		std::ostringstream ss;
+		auto killers = getKillers();
+		std::map<std::string, uint16_t> names;
+		for (auto& killer : killers) {
+			std::string killerName = killer->getName();
+			if (names.find(killerName) != names.end()) {
+				names[killerName]++;
+			} else {
+				names[killerName] = 1;
+			}
+		}
+
+		size_t countNames = names.size();
 		if (lastHitCreature) {
-			ss << "You recognize " << getNameDescription() << ". " << (getSex() == PLAYERSEX_FEMALE ? "She" : "He") << " was killed by " << lastHitCreature->getNameDescription() << '.';
+			if (countNames == 1) {
+				ss << "You recognize " << getNameDescription() << ". " << (getSex() == PLAYERSEX_FEMALE ? "She" : "He") << " was killed by " << lastHitCreature->getNameDescription() << '.';
+			} else if (mostDamageCreature && names[mostDamageCreature->getName()] >= 1) {
+				ss << "You recognize " << getNameDescription() << ". " << (getSex() == PLAYERSEX_FEMALE ? "She" : "He") << " was killed by " << mostDamageCreature->getNameDescription();
+				if (lastHitCreature != mostDamageCreature && names[lastHitCreature->getName()] == 1) {
+					ss << " and " << lastHitCreature->getNameDescription();
+					if (countNames > 2) {
+						ss << " and others.";
+					}
+				} else {
+					ss << " and others.";
+				}
+			}
 		} else {
 			ss << "You recognize " << getNameDescription() << '.';
```

Comment:
maybe the "He/She was killed by something evil." should be added here? 🤔

Reply:
It cannot be done, because we are not taking into account the damage received by evil sources, I am still thinking of a better idea, thanks for reviewing, let's continue here to find a better way to implement evil sources and thus have this completely, the I really would like to have this completely on my server